rfd/0128-aws-integration-tests.md

```diff
@@ -0,0 +1,191 @@
+---
+authors: Tiago Silva (tiago.silva@goteleport.com)
+state: draft
+---
+
+# RFD 128 - AWS End To End Tests
+
+## Required Approvers
+
+- Engineering: `@r0mant && @smallinsky`
+- Security: `@reedloden || @jentfoo`
+
+## What
+
+As part of the increase of the reliability of Teleport, we want to add integration
+tests for AWS. This will allow us to test the integration with AWS without having
+to rely on manual testing.
+
+## Why
+
+We want to increase the reliability of Teleport by adding integration tests for
+AWS. This will allow us to test the integration with AWS without having to rely
+on manual testing. This will also allow us to test the integration with AWS in
+a more reliable process to ensure we don't introduce regressions in the future
+or if we do, we can catch them early.
+
+Teleport integrates deeply with AWS to provide a seamless experience for users
+when using auto-discovery of nodes, databases and Kubernetes clusters. This
+integration is critical for the success of Teleport and we want to ensure that
+we can test it reliably.
+
+Each integration test will use the minimum required
+AWS API permissions to test the integration. This will ensure that we don't
+introduce regressions by changing the permissions required by Teleport but those
+changes are not detected because another test requires the same permissions.
+
+The goal of this RFD is to define the scope of the integration tests for AWS
+and how we will run them.
+
+## When
+
+AWS integration tests will be added to the Teleport CI pipeline as part of the
+cronjob that triggers integrations each day. During the release process, the
```

Comment:
> During the release process, the integration tests will be run automatically as part of the release process.

I'm not sure about this tbh. I understand the intent, but IMO blocking a release on the integration tests failure is suboptimal. I think that if something fails during release process, the issue was detected too late, and also imagine we need to push an important release we can't delay or something.

I know we originally discussed running these nightly but thinking some more it would make sense to see if we can run them on each PR, it's always better to have fast feedback loop. I think it should be doable if we design the tests themselves to they don't conflict.

WDYT?

Reply:
I was wondering if we could use AWS CloudFormation to create the testing environment 🤔 
We deploy everything we need there: EKS, EC2, DBs, ... and wait for it to complete
We then start running the tests against those resources

Something like this
https://cluster-api-aws.sigs.k8s.io/development/e2e.html
https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/6705ad80eb1c3fe47036505485f3c24141e88a59/Makefile#L407

Creating those resources is a slow process for most (we know that EKS alone takes around 20 minutes 😭 ).
So, I'm not sure doing this for every PR is feasible
I would suggest we start with a nightly job and only for `master` and `branch/vX` and an on-demand way for running it against a specific commit or branch

Reply:
@r0mant I am ok running tests for each PR but it is a bit less secure since we allow executing E2E for code that wasn't properly reviewed yet. 

To improve security, if the contributor is external, we only allow E2E to run if the PR is approved.